x-pack/platform/plugins/shared/metrics_experience/server/lib/get_metric_fields.ts

```diff
@@ -0,0 +1,168 @@
+import { getEcsFieldDescriptions } from './get_ecs_field_descriptions';
+import { extractTimeSeriesFields } from './extract_time_series_fields';
+import { batchedSampleMetricDocuments } from './sample_metric_documents';
+import { extractDimensions } from './extract_dimensions';
+import { buildMetricField } from './build_data_stream_field';
+import { retrieveFieldCaps } from './retrieve_fieldcaps';
+
+export async function getMetricFields({
+  indexPattern,
+  fields = '*',
+  from,
+  to,
+  esClient,
+  logger,
+}: {
+  esClient: ElasticsearchClient;
+  indexPattern: string;
+  fields?: Fields;
+  from?: string;
+  to?: string;
+  logger: Logger;
+}): Promise<MetricField[]> {
+  if (!indexPattern) return [];
+
+  try {
+    // Wait for all field caps requests to complete
+    const fieldCapsResults = await retrieveFieldCaps({
+      esClient,
+      indexPattern,
+      to,
+      from,
+      fields,
+    });
+
+    // Process results from each data stream
+    const allFields: MetricField[] = [];
+    const allSamplingPromises: Promise<Map<string, string[]>>[] = [];
+
+    for (const { dataStreamName, fieldCaps } of fieldCapsResults) {
+      if (Object.keys(fieldCaps).length === 0) continue;
+
+      const timeSeriesFields = extractTimeSeriesFields(fieldCaps);
+
+      // Get all dimensions for context
+      const allDimensions = extractDimensions(fieldCaps);
+
+      // Process only metric fields (not dimensions) for the main results
+      const metricFields = timeSeriesFields.filter((field) => field.fieldType === 'metric');
+
+      // Build initial data stream objects for deduplication
+      const initialFields: MetricField[] = [];
+      for (const { fieldName, type, typeInfo } of metricFields) {
+        initialFields.push(
+          buildMetricField(
+            fieldName,
+            dataStreamName, // Use data stream name as index value
+            allDimensions, // Use all dimensions initially
+            type,
+            typeInfo
+          )
+        );
+      }
+
+      // Deduplicate fields before sampling to reduce work
+      const deduped = deduplicateFields(initialFields);
+
+      // Collect metric names for sampling
+      const metricNames = deduped.map((field) => field.name);
+
+      if (metricNames.length > 0) {
+        // Add sampling promise for this data stream
+        allSamplingPromises.push(
+          batchedSampleMetricDocuments(esClient, dataStreamName, metricNames, 500, logger)
+        );
+
+        // Store the fields with their field caps for later processing
+        for (const field of deduped) {
+          (field as any)._fieldCaps = fieldCaps;
+        }
+
+        allFields.push(...deduped);
+      }
+    }
+
+    // Wait for all sampling operations to complete
+    const samplingResults = await Promise.all(allSamplingPromises);
+
+    // Merge sampling results
+    const allSampledDocs = new Map<string, string[]>();
+
+    samplingResults.forEach((samplingMap) => {
```

Comment:
nit: we can continue using `for ...of` like in the other places for consistency and also for better performance